File: titanium-rdf-api/src/main/scala/eu/ostrzyciel/jelly/convert/titanium/TitaniumJellyEncoderImpl.scala
package eu.ostrzyciel.jelly.convert.titanium

import com.apicatalog.rdf.api.{RdfConsumerException, RdfQuadConsumer}
import eu.ostrzyciel.jelly.convert.titanium.Constants.DT_STRING
import eu.ostrzyciel.jelly.convert.titanium.internal.TitaniumConverterFactory
import eu.ostrzyciel.jelly.convert.titanium.internal.TitaniumRdf.*
import eu.ostrzyciel.jelly.core.RdfProtoSerializationError
import eu.ostrzyciel.jelly.core.ProtoEncoder
import eu.ostrzyciel.jelly.core.proto.v1.*

import scala.collection.immutable
import scala.collection.mutable.ListBuffer
import scala.jdk.CollectionConverters.*

private class TitaniumJellyEncoderImpl(options: RdfStreamOptions) extends TitaniumJellyEncoder:
  private val buffer: ListBuffer[RdfStreamRow] = new ListBuffer[RdfStreamRow]()
  private val encoder = TitaniumConverterFactory.encoder(ProtoEncoder.Params(
    // We set the stream type to QUADS, as this is the only type supported by Titanium.
    options = options.copy(
      physicalType = PhysicalStreamType.QUADS,
      logicalType = LogicalStreamType.FLAT_QUADS,
      // It's impossible to emit generalized statements or RDF-star in Titanium.
      generalizedStatements = false,
      rdfStar = false,
    ),
    enableNamespaceDeclarations = false,
    maybeRowBuffer = Some(buffer),
  ))

  override def quad(
    subject: String, 
    predicate: String,
    `object`: String, 
    datatype: String, 
    language: String, 
    direction: String,
    graph: String
  ): RdfQuadConsumer =
    // IRIs and bnodes don't need further processing. For literals, we must allocate
    // intermediate objects.
    try {
      if RdfQuadConsumer.isLiteral(datatype, language, direction) then
        val literal = if RdfQuadConsumer.isLangString(datatype, language, direction) then
          LangLiteral(`object`, language)
        else if datatype == DT_STRING then
          SimpleLiteral(`object`)
        else
          DtLiteral(`object`, datatype)
        encoder.addQuadStatement(subject, predicate, literal, graph)
      else
        encoder.addQuadStatement(subject, predicate, `object`, graph)
    } catch {
      case e: RdfProtoSerializationError => throw new RdfConsumerException(e.getMessage, e)
    }
    this

  final override def getRowCount: Int = buffer.size

  final override def getRowsScala: Seq[RdfStreamRow] =
    val list = buffer.toList
    buffer.clear()
    list

  final override def getRowsJava: java.lang.Iterable[RdfStreamRow] = getRowsScala.asJava

  final override def getOptions: RdfStreamOptions = encoder.options

File: titanium-rdf-api-java/src/main/java/eu/neverblink/jelly/convert/titanium/TitaniumJellyEncoderImpl.java
package eu.neverblink.jelly.convert.titanium;

import static eu.neverblink.jelly.convert.titanium.TitaniumConstants.DT_STRING;

import com.apicatalog.rdf.api.RdfConsumerException;
import com.apicatalog.rdf.api.RdfQuadConsumer;
import eu.neverblink.jelly.convert.titanium.internal.TitaniumConverterFactory;
import eu.neverblink.jelly.convert.titanium.internal.TitaniumLiteral;
import eu.neverblink.jelly.core.InternalApi;
import eu.neverblink.jelly.core.ProtoEncoder;
import eu.neverblink.jelly.core.RdfProtoSerializationError;
import eu.neverblink.jelly.core.proto.v1.LogicalStreamType;
import eu.neverblink.jelly.core.proto.v1.PhysicalStreamType;
import eu.neverblink.jelly.core.proto.v1.RdfStreamOptions;
import eu.neverblink.jelly.core.proto.v1.RdfStreamRow;
import java.util.ArrayList;
import java.util.Collection;

@InternalApi
final class TitaniumJellyEncoderImpl implements TitaniumJellyEncoder {

    private final ProtoEncoder<Object> encoder;

    private final Collection<RdfStreamRow> buffer = new ArrayList<>();

    public TitaniumJellyEncoderImpl(RdfStreamOptions options) {
        // We set the stream type to QUADS, as this is the only type supported by Titanium.
        final var supportedOptions = options
            .clone()
            .setPhysicalType(PhysicalStreamType.QUADS)
            .setLogicalType(LogicalStreamType.FLAT_QUADS)
            // It's impossible to emit generalized statements or RDF-star in Titanium.
            .setGeneralizedStatements(false)
            .setRdfStar(false);

        this.encoder = TitaniumConverterFactory.getInstance()
            .encoder(ProtoEncoder.Params.of(supportedOptions, false, buffer));
    }

    @Override
    public int getRowCount() {
        return buffer.size();
    }

    @Override
    public Iterable<RdfStreamRow> getRows() {
        return buffer;
    }

    @Override
    public void clearRows() {
        buffer.clear();
    }

    @Override
    public RdfStreamOptions getOptions() {
        return encoder.getOptions();
    }

    @Override
    public RdfQuadConsumer quad(
        String subject,
        String predicate,
        String object,
        String datatype,
        String language,
        String direction,
        String graph
    ) throws RdfConsumerException {
        // IRIs and bnodes don't need further processing. For literals, we must allocate
        // intermediate objects.
        try {
            if (RdfQuadConsumer.isLiteral(datatype, language, direction)) {
                final TitaniumLiteral literal;
                if (RdfQuadConsumer.isLangString(datatype, language, direction)) {
                    literal = new TitaniumLiteral.LangLiteral(object, language);
                } else if (datatype.equals(DT_STRING)) {
                    literal = new TitaniumLiteral.SimpleLiteral(object);
                } else {
                    literal = new TitaniumLiteral.DtLiteral(object, datatype);
                }

                encoder.handleQuad(subject, predicate, literal, graph);
            } else {
                encoder.handleQuad(subject, predicate, object, graph);
            }
        } catch (RdfProtoSerializationError e) {
            throw new RdfConsumerException(e.getMessage(), e);
        }

        return this;
    }
}
